src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,508 @@
+}
+
+function subBoxIndex(idx: number, offset: vec3, cropSize: vec3, size: vec3) {
+  return idx % cropSize[0] + offset[0] +
+      (Math.floor(idx / cropSize[0]) % cropSize[1] + offset[1]) * size[0] +
+      (Math.floor(idx / (cropSize[0] * cropSize[1])) % cropSize[2] + offset[2]) * size[0] *
+      size[1] +
+      (Math.floor(idx / (cropSize[0] * cropSize[1] * cropSize[2]))) * size[0] * size[1] * size[2];
+}
+
+export function copyBufferOverlap(
+    sourceCorner: vec3, sourceSize: vec3, sourceView: ArrayType, destCorner: vec3, destSize: vec3,
+    destView: ArrayType, dataType: DataType) {
+  // UINT64 data is packed two-at-a-time into a UINT32 array, so we handle it as a special case.
+  let copyFunction = dataType === DataType.UINT64 ? (j: number, k: number) => {
+    destView[2 * k] = sourceView[2 * j];
+    destView[2 * k + 1] = sourceView[2 * j + 1];
+  } : (j: number, k: number) => {
+    destView[k] = sourceView[j];
+  };
+
+  // Global Coordinates
+  const commonLower = vec3.max(vec3.create(), sourceCorner, destCorner);
+  const sourceUpper = vec3.add(vec3.create(), sourceCorner, sourceSize);
+  const destUpper = vec3.add(vec3.create(), destCorner, destSize);
+  const commonUpper = vec3.min(vec3.create(), sourceUpper, destUpper);
+  const commonSize = vec3.subtract(vec3.create(), commonUpper, commonLower);
+
+  const sourceLower = vec3.subtract(vec3.create(), commonLower, sourceCorner);
+  const destLower = vec3.subtract(vec3.create(), commonLower, destCorner);
+
+
+  for (let i = 0; i < prod(commonSize); ++i) {
+    const j = subBoxIndex(i, /*offset=*/sourceLower, /*cropSize=*/commonSize, /*size=*/sourceSize);
+    const k = subBoxIndex(i, /*offset=*/destLower, /*cropSize=*/commonSize, /*size=*/destSize);
+    copyFunction(j, k);
+  }
+}
+
+// Computes a consistent key string from a chunk grid position.
+//
+// It's tempting to use chunk.key, in particular because these values will
+// often be the same, but we won't always have access to a fully-specified
+// chunk, and there's no contractual guarantee that its key will be equal to
+// the value returned here.
+function gridPositionKey(gridPosition: vec3) {
+  return gridPosition.toLocaleString();
+}
+
+// In addition to acting as a VolumeChunk for the purposes of a ChunkManager
+// object, also performs the book-keeping necessary to prepare the data buffer
+// used as input by the computation that provides its data. This includes
+// fetching chunk data from other datasources.
+export class ComputedVolumeChunk extends VolumeChunk implements ChunkStateListener {
+  private computationParams_?: ComputationParameters;
+
+  private cancellationToken_?: CancellationToken;
+  private resolve_?: () => void;
+  private reject_?: (reason: any) => void;
+  private success_ = false;
+  private failure_?: any;
+  private computing_ = false;
+
+  private outputSource_?: ComputedVolumeChunkSource;
+  private originGridPositions_ = new Map<String, vec3>();
```

Comment:
It would be good to document the key/value meaning.

Reply:
More generally it would be quite helpful to document the meaning/invariants of these members --- I know most of the code in Neuroglancer currently doesn't set a great standard in that regard, but I think it would be particularly helpful here.

Reply:
Comments added generally. If there are any areas of the code that need more explanation, I'm happy to add it.